Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -171,79 +168,164 @@ private void DiscardReagents(ChemMasterComponent chemMaster, string reagentId, F
+            if (maybeContainer is not { Valid: true } container
+                || !_solutionContainerSystem.TryGetSolution(
+                    container, SharedChemMaster.BottleSolutionName, out var solution))
+            {
+                return; // output can't fit reagents
+            }
+
             // Ensure the amount is valid.
-            if (message.Amount == 0 || message.Amount > chemMaster.BottleProductionLimit)
-                return;
+            if (message.Amount == 0 || message.Amount > solution.AvailableVolume) return;
+            
+            if (!WithdrawFromBuffer(chemMaster, message.Amount, user, out var withdrawal)) return;
+
+            Label(container, message.Label);
+            DebugTools.Assert(_solutionContainerSystem.TryAddSolution(
+                container, solution, withdrawal));
 
-            CreatePillsOrBottles(chemMaster, pills: false, message.Amount, message.Label, message.Session.AttachedEntity);
             UpdateUiState(chemMaster);
             ClickSound(chemMaster);
         }
 
-        private void CreatePillsOrBottles(ChemMasterComponent chemMaster, bool pills, FixedPoint2 amount, string label, EntityUid? user)
+        private bool WithdrawFromBuffer(
+            IComponent chemMaster,
+            FixedPoint2 neededVolume, EntityUid? user,
+            [NotNullWhen(returnValue: true)] out Solution? outputSolution)
         {
-            if (!_solutionContainerSystem.TryGetSolution(chemMaster.Owner, SharedChemMaster.BufferSolutionName, out var bufferSolution))
-                return;
+            outputSolution = null;
+            
+            if (!_solutionContainerSystem.TryGetSolution(
+                    chemMaster.Owner, SharedChemMaster.BufferSolutionName, out var solution))
+            {
+                return false;
+            }
 
             var filter = user.HasValue ? Filter.Entities(user.Value) : Filter.Empty();
-            if (bufferSolution.TotalVolume == 0)
+            if (solution.TotalVolume == 0)
             {
                 _popupSystem.PopupCursor(Loc.GetString("chem-master-window-buffer-empty-text"), filter);
-                return;
+                return false;
             }
 
-            var individualVolume = FixedPoint2.Min(bufferSolution.TotalVolume / amount, FixedPoint2.New(pills ? 50 : 30));
-            if (individualVolume < FixedPoint2.New(1))
+            // ReSharper disable once InvertIf
+            if (neededVolume > solution.CurrentVolume)
             {
                 _popupSystem.PopupCursor(Loc.GetString("chem-master-window-buffer-low-text"), filter);
-                return;
+                return false;
             }
 
-            for (int i = 0; i < amount; i++)
+            outputSolution = solution.SplitSolution(neededVolume);
+            return true;
+        }
+
+        private void Label(EntityUid ent, string label)
```

Comment:
should be replaced with a public API into LabelSystem but uhhh when somebody codes it